### DISCUSSION TITLE
Comparing generic types

### DISCUSSION THREAD

**Discussion by NSExceptional** [Azoy/Echo#44] [Q&A] [answered]
Right now, I need to check if a type is Swift's `Array` type. `Array` is generic though. Is there a way I can compare the root type of a generic type for equality like that?

**Comment by Azoy** [comment, marked as answer]
The simplest answer I can give is to compare the context descriptors (because there will only ever be 1 of them per type).
```swift
let array = reflectStruct([Int].self)!.descriptor

// Assuming someType refers to some variable containing `TypeMetadata`
guard someType.contextDescriptor.ptr == array.ptr else {
  // The outer type of someType is not an Array
  return nil
}
```

**Reply by Azoy** [reply]
The Equatable conformances for these descriptors are defined on the concrete type itself (vs. the base `ContextDescriptor` protocol) to be able to return values of `ContextDescriptor`. Hence the need to compare the pointers themselves vs. the type.

**Reply by NSExceptional** [reply]
Ah yes, comparing the `[context]Descriptor` properties is what I needed to do :) Thanks

```swift
func testArrayMetadataCheck() {
    let anyarray = reflectStruct([Any].self)!
    let intarray = reflectStruct([Int].self)!
    
    XCTAssert(anyarray.descriptor == intarray.descriptor)
}
```
